book/src/dev/rfcs/0002-parallel-verification.md

```diff
@@ -0,0 +1,324 @@
+    processing the checkpoint range. (This constraint is an implicit part of
+    the `CheckpointVerifier` design.)
+  * `process_checkpoint_range`: makes sure that the blocks in the checkpoint
+    range have an unbroken chain of previous block hashes.
+* **State Updates:**
+  * *As Above:* the `CheckpointVerifier` returns success to the `ChainVerifier`,
+    which sends verified `Block`s to the state service.
+  * The state service implements the chain order consensus rule, which makes sure
+    each block is added to state after its previous block.
+
+### Block Verification
+[block-verification]: #block-verification
+
+The `BlockVerifier` performs detailed verification of recent blocks, in parallel.
+
+Here is how the `BlockVerifier` implements each verification stage:
+
+* **Structural Verification:**
+  * *As Above:* the `BlockVerifier` accepts parsed `Block` structs.
+* **Semantic Verification:**
+  * *As Above:* verifies each field in the block. Defers any data dependencies as
+    long as possible, awaits those data dependencies, then performs data
+    dependent checks.
+  * Note: Since futures are executed concurrently, we can use the same function
+    to:
+    * perform context-free verification,
+    * perform verification with deferred data dependencies,
+    * await data dependencies, and
+    * check data dependencies.
+    To maximise concurrency, we should write verification functions in this
+    specific order, so the awaits are as late as possible.
+* **State Updates:**
+  * *As Above:* the `BlockVerifier` returns success to the `ChainVerifier`,
+    which sends verified `Block`s to the state service.
+  * The state service implements the chain order consensus rule, which makes sure
+    each block is added to state after its previous block.
+
+## Zcash Protocol Design
+[zcash-protocol]: #zcash-protocol
+
+When designing a change to the Zcash protocol, minimise the data dependencies
+between blocks.
+
+Try to create designs that:
+* Eliminate data dependencies,
+* Make the changes depend on a version field in the block header or transaction,
+* Make the changes depend on the current Network Upgrade,
+* Make the changes depend on a field in the current block, with an additional
+  consensus rule to check that field against previous blocks, or
+* Prefer dependencies on older blocks, rather than newer blocks
+  (older blocks are more likely to verify earlier).
+
+Older dependencies have a design tradeoff:
+* depending on multiple blocks is complex,
+* depending on the previous block makes parallel verification harder.
+
+When making decisions about this dependency tradeoff, consider:
+* how the data dependency could be deferred, and
+* the CPU cost of the verification - if it is trivial, then it does not matter if
+  the verification is parallelised.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+**TODO:** this design is a bit complicated, but we think it's necessary to
+          achieve our goals.
```

Comment:
I don't think it's complicated tbh, it seems fairly natural and aligned with a futures based application.

Reply:
Getting here was complicated though 😅

And it's definitely more complicated than verifying blocks in strict order.

I'm happy to adjust the wording here, do you have any suggestions?

Reply:
no changes necessary.